fastlane_bot/modes/triangle_multi.py

```diff
@@ -22,65 +20,141 @@ class ArbitrageFinderTriangleMulti(ArbitrageFinderTriangleBase):
-        """
-        see base.py
-        """
-
-        if self.base_exchange != "carbon_v1":
-            raise ValueError("base_exchange must be carbon_v1 for `multi` mode")
-
-        if candidates is None:
-            candidates = []
-
-        combos = self.get_combos(self.flashloan_tokens, self.CCm, arb_mode=self.arb_mode)
-
-        for src_token, miniverse in combos:
-            try:
-                CC_cc = CPCContainer(miniverse)
-                O = MargPOptimizer(CC_cc)
-                pstart = self.build_pstart(CC_cc, CC_cc.tokens(), src_token)
-                r = O.optimize(src_token, params=dict(pstart=pstart))
-                trade_instructions_dic = r.trade_instructions(O.TIF_DICTS)
-                if trade_instructions_dic is None or len(trade_instructions_dic) < 3:
-                    # Failed to converge
-                    continue
-                trade_instructions_df = r.trade_instructions(O.TIF_DFAGGR)
-                trade_instructions = r.trade_instructions()
-
-            except Exception as e:
-                self.ConfigObj.logger.info(f"[triangle multi] {e}")
-                continue
-            profit_src = -r.result
+    def assert_supported(self):
+        assert self.base_exchange == "carbon_v1", "base_exchange for `multi_triangle` mode must be `carbon_v1`"
 
-            # Get the cids
-            cids = [ti["cid"] for ti in trade_instructions_dic]
-
-            # Calculate the profit
-            profit = self.calculate_profit(src_token, profit_src, self.CCm, cids)
-            if str(profit) == "nan":
-                self.ConfigObj.logger.debug("profit is nan, skipping")
-                continue
+    def get_combos(self, flashloan_tokens: List[str], CCm: Any, arb_mode: str) -> Tuple[List[str], List[Any]]:
+        """
+        Get combos for triangular arbitrage
+
+        Parameters
+        ----------
+        flashloan_tokens : list
+            List of flashloan tokens
+        CCm : object
+            CCm object
+        arb_mode : str
+            Arbitrage mode
+
+        Returns
+        -------
+        combos : list
+            List of combos
 
-            # Handle candidates based on conditions
-            candidates += self.handle_candidates(
-                best_profit,
-                profit,
-                trade_instructions_df,
-                trade_instructions_dic,
-                src_token,
-                trade_instructions,
-            )
+        """
+        combos = []
+        if arb_mode in ["b3_two_hop"]:
+            combos = [
+                (tkn0, tkn1)
+                for tkn0, tkn1 in product(flashloan_tokens, flashloan_tokens)
+                # note that the pair is tkn0/tkn1, ie tkn1 is the quote token
+                if tkn0 != tkn1
+            ]
+        else:
+            all_base_exchange_curves = CCm.byparams(exchange=self.base_exchange).curves
```

Comment:
comments on what you are doing and why you are doing it would be helpful

Reply:
Outside the scope of this PR (same code used prior to it)